### PR TITLE
[Omega][tools/depends][target] Bump curl 8.10.0

### PR DESCRIPTION
**Pull request by blunden**
## Description

Backport of https://github.com/xbmc/xbmc/pull/25734

Bumps curl to 8.10.0. Backported to Omega, already merged in master.

## Motivation and context

Should fix issues with FTPES sources (https://github.com/xbmc/xbmc/issues/25118) as well as patch a few vulnerabilities. FTPES sources have been broken ever since the first Kodi Omega release and is confirmed to have been fixed in curl version 8.9.1. I simply picked the latest version.

https://forum.kodi.tv/showthread.php?tid=377271
https://forum.libreelec.tv/thread/28779-sftp-broken-in-le12/?postID=194487#post194487
https://curl.se/docs/vulnerabilities.html

## How has this been tested?

Tested on Android ARM 64-bit.

## What is the effect on users?

Besides fixing the issue mentioned above, it should hopefully not have many visible changes.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by blunden** [comment]
I unfortunately don't know what's up with those nghttp2 errors specifically on Apple platforms. Not sure if parts of the patch below is still needed:

https://github.com/blunden/xbmc/commit/030e62a663ecf1d92e6eab160fd18fd332a1c43d

I'm sure someone who has looked into this before might recognize the issue. 🙂

**Comment by neo1973** [comment]
Jenkins build this please

**Comment by neo1973** [comment]
It probably builds on master because of #25115. Not sure how to deal with that.

**Comment by blunden** [comment]
@neo1973 How would you normally deal with this. Just leaving features broken forever in Omega doesn't seem like a good option. Backporting those build changes too (but removing the already upstreamed patch) is a bigger change than might usually be expected for a backport though, I assume?

**Comment by sundermann** [comment]
What about adding `--without-nghttp2`? Could work unless we need the `nghttp2` features

**Comment by blunden** [comment]
That might work. We would be disabling HTTP/2 entirely in that case though as far as I understand it. Perhaps people are using HTTP/2 for HTTP/HTTPS content streaming?

**Comment by blunden** [comment]
> > How has this been tested?
> > This time not at all.
> 
> All backports must be tested, especially if are high risk of broke things (as is this case).

Yes, that's true. I had a version mismatch build issue for flatbuffers due to previously building master, and I wanted to push something as discussion material while I had some attention around this again after a month of radio silence. That's why I was transparent about not having tested it. 🙂 Maybe it should've been marked as WIP though.

I also knew that several other people have been running newer curl versions on Omega.

In this case, even if I had tested it on the platforms I use, that still wouldn't have caught the build issues on OSX/iOS/TvOS.

> All backports should be the same code (cherry pick) that has been merged into master or with minimal changes due to merge conflicts. Having to change the code is a clear symptom of a high risk backport.

I agree. It's not identical simply because the nghttp2 patch that I was previously told to remove doesn't exist in Omega, which I found after trying to git am a .patch file of the master change (a work flow I find easier than git cherry-pick, especially across repos). Now we know the PR that patch was a part of might actually be necessary to build it on Apple's platforms. 🙁

I would welcome suggestions on how to proceed. 🙂

**Comment by icarok99** [comment]
> > > How has this been tested?
> > > This time not at all.
> > 
> > 
> > All backports must be tested, especially if are high risk of broke things (as is this case).
> 
> Yes, that's true. I had a version mismatch build issue for flatbuffers due to previously building master, and I wanted to push something as discussion material while I had some attention around this again after a month of radio silence. That's why I was transparent about not having tested it. 🙂 Maybe it should've been marked as WIP though.
> 
> I also knew that several other people have been running newer curl versions on Omega.
> 
> In this case, even if I had tested it on the platforms I use, that still wouldn't have caught the build issues on OSX/iOS/TvOS.
> 
> > All backports should be the same code (cherry pick) that has been merged into master or with minimal changes due to merge conflicts. Having to change the code is a clear symptom of a high risk backport.
> 
> I agree. It's not identical simply because the nghttp2 patch that I was previously told to remove doesn't exist in Omega, which I found after trying to git am a .patch file of the master change (a work flow I find easier than git cherry-pick, especially across repos). Now we know the PR that patch was a part of might actually be necessary to build it on Apple's platforms. 🙁
> 
> I would welcome suggestions on how to proceed. 🙂

The omega version modules are out of outdated, This is the real reason you can't find nghttp2.

**Comment by blunden** [comment]
Any suggestions?

**Comment by garbear** [comment]
Well, we need to fix FTPES, and curl 8.10.0 does that, so somehow we need to get this into Omega.

We're caught in a tricky situation, because we could either backport a large amount of CMake work with low effort and risk breaking something, or redo the 8.10.0 bump from scratch against Omega with high effort, but little risk of breaking something.

I have a stake in this game, because I've backported all the CMake work that @fuzzard did to my retroplayer branch based on Omega. My open PRs kept breaking, so I backported everything that resolved a conflict. You can see the PRs I do builds with by scrolling back here: https://github.com/garbear/xbmc/commits/retroplayer-21.1

Specifically, these are all the PRs I test with RetroPlayer builds:

```
PR 21286
PR 22652
PR 23944
PR 24700
PR 24719
PR 24709
PR 24755
PR 24064
PR 24508
PR 24605
PR 24821
PR 24837
PR 24841
PR 24860
PR 24898
PR 24910
PR 24924
PR 24929
PR 24931
PR 24938
PR 24939
PR 24940
PR 24967
PR 24981
PR 25008
PR 25009
PR 25031
PR 25033
PR 22919
PR 25039
PR 25029
PR 25049
PR 25071
PR 25073
PR 25077
PR 25078
PR 25115
PR 25139
PR 25141
PR 25144
PR 25170
PR 25178
PR 25179
PR 25180
PR 25193
PR 25197
PR 25207
PR 25228
PR 24623
PR 25230
PR 25276
PR 25286
PR 25326
PR 25296
PR 25310
PR 25245
PR 25300
PR 25370
PR 25390
PR 25401
PR 25419
PR 25428
PR 25435
PR 25444
PR 25456
PR 25466
PR 25468
PR 25476
PR 25479
PR 25509
PR 25512
PR 25514
PR 25584
PR 25598
PR 25610
PR 25614
PR 25623
PR 25680
PR 25690
PR 25691
PR 25701
PR 25889
PR 25897
```

If you can get the 8.10.0 patch to apply with any of those PRs, I can vouch that my builds haven't uncovered any issues with the backport.

So I think the low-effort way of "backport just enough" is the way to go. But others can chime in on the risk.

**Comment by fuzzard** [comment]
Just disable nghttp2 in curl. move on.

**Comment by fuzzard** [comment]
You will need https://github.com/fuzzard/xbmc/commit/3b4ef4bc98ed84a3fe8be7709fc056ec02789229

curl has moved to using Requires and Requires.private in their pkgconfig file. As apple uses system zlib, there is no pc file, and therefore any pkgconfig searches for libcurl error. This causes our cmake code to have empty PC_CURL_STATIC_LDFLAGS, and therefore the explicit linking of nghttp isnt added, causing the link failures later.

This is an Omega only patch.

**Comment by blunden** [comment]
Awaiting the result of the build bot before even attempting to test this since I have no good way to build for Apple platforms at the moment.

File: tools/depends/target/curl/Makefile [hunk 1]
```diff
@@ -22,6 +22,7 @@ CONFIGURE=cp -f $(CONFIG_SUB) $(CONFIG_GUESS) .; \
                       --without-gsasl \
                       --without-hyper \
                       --without-ngtcp2 \
+                      --without-nghttp2 \
                       --without-nghttp3 \
```
Comment (fuzzard) [inline tools/depends/target/curl/Makefile:26]: You don't need to disable nghttp2 with the patch

Comment (blunden) [reply tools/depends/target/curl/Makefile:26]: Oh, I see. Should I just git reset and force push it out then? :)

Comment (fuzzard) [reply tools/depends/target/curl/Makefile:26]: Yeah,

**Comment by blunden** [comment]
Hold on this I get a chance to test it. 🙂

**Comment by blunden** [comment]
Tested working so far on Android (ARM64).

I initially had some weird problems that I'm now fairly sure were to do with the fact that I mistakenly used an old subdomain for my server where the A record (IPv4 address) was correct but the AAAA record (IPv6 address) pointed to the wrong server. This seemed to confuse either Kodi or curl as it seemed to correctly fall back to the IPv4 address after getting a connection refused on the IPv6 address, but then it cached the IPv6 address and seemingly only tried (and) failed to use that according to the kodi log. The result was that it would connect fine once and let you browse around the server but then fail any subsequent connections (such as streaming a video, disconnecting and then trying to reconnect again, etc.).

I welcome other testers as well. 🙂

**Review by thexai — APPROVED** [review]

**Review by neo1973 — APPROVED** [review]

**Comment by neo1973** [comment]
Thank you @blunden and everyone involved in backporting this :+1:

**Pull request merged**